### PR TITLE
Fix VM related test cases

### PR DESCRIPTION
**Pull request by torchiaf**
Related issue https://github.com/harvester/tests/issues/1196

The VM tests are failing because the node `harvester-node-0` has a  `test-custom-name` customName in Harvester configuration; This breaks the cypress tests in matching the correct UI elements.

image

- I've updated the tests to support host's `customName`.
- I added the `customName` in host config. **This will require to update the config in Jenkins pipelines**:
  - `host` should be always an array
  - if `customName` is empty, the `name` will be used during the tests

```
  "host": [
    {
      "name": "",
      "customName": "",
      "disks": [
        {
          "name": "",
          "devPath": ""
        }
      ],
      "witnessNode": false
    }
  ],
```

### CODE REVIEW

File: cypress/utils/utils.ts [hunk 1]
```diff
@@ -46,3 +46,15 @@ export function base64DecodeToBuffer(string: string) {
 export function base64Decode(string: string) {
   return !string ? string : base64DecodeToBuffer(string.replace(/[-_]/g, char => char === '-' ? '+' : '/')).toString();
 }
+
+export const nodes = {
+  filterWitnessNode: (hosts: {name: string, witnessNode: boolean}[]) => {
```
Comment (a110605) [inline cypress/utils/utils.ts:51]: `filterWitnessNode()` and `witnessNode` config feild are not used in the code. 
Are they prepare for future test cases use?

File: cypress/testcases/virtualmachines/scheduling.spec.ts [hunk 2]
```diff
@@ -14,25 +14,27 @@ describe('VM scheduling on Specific node', () => {
 
   it('Schedule VM on the Node which is Enable Maintenance Mode', () => {
     const hostList = Cypress.env('host');
-    const hostNames: string[] = hostList.map((host: any) => host.name);
-    const maintenanceNode = hostNames[0]
-    const filterMaintenanceNames = hostNames.filter(name => name !== maintenanceNode);
+
+    const hostNames: string[] = hostList.map((node: Node) => node.customName || node.name);
+
+    const maintenanceNodeName = hostNames[0]
```
Comment (a110605) [inline cypress/testcases/virtualmachines/scheduling.spec.ts:20]: Even host is prefilled with a item in `cypress.env.json`.
How about add an early return error message if hostList is empty. 
Same in node-scheduling.spec.ts.

File: cypress/utils/utils.ts (hunk 1 above)
Comment (torchiaf) [reply cypress/utils/utils.ts:51]: Yes, I would keep this code for future test cases.

File: cypress/testcases/virtualmachines/scheduling.spec.ts (hunk 2 above)
Comment (torchiaf) [reply cypress/testcases/virtualmachines/scheduling.spec.ts:22]: Good idea, I just added a hosts util to do that.

**Review by a110605 — APPROVED** [review]
Code change LTGM.
Could you verify test cases should pass against jenkins pipeline env again before merge?

**Comment by TachunLin** [comment]
Tested on the main Jenkins for this test changes along with the harvester-baremetal-ansible pull request 
* Added the corresponding witness mode setting in `cypress.config.ts.j2` and `cypress.env.json.j2`
  https://github.com/TachunLin/harvester-baremetal-ansible/tree/fix-cypress-test-1319

According to the latest test result on the staging Jenkins of:
* Jobs: http://172.19.98.192:8080/job/harvester-cypress-test/8/console
* Test report: https://minio.provo.rancherlabs.com:31524/cypress-test-report/cypress/results/20240617-101755-512c1bd/index.html


image

It will encounter the `Cannot read properties of undefined (reading split)` failure, just as what we encounter last week. 

Checked the `package.json`, the adm-zip version seems not changed, it looks like it should be rebase to sync with the latest main. 
https://github.com/torchiaf/harvester-tests/blob/cypress-1196/cypress/package.json

**Comment by torchiaf** [comment]
Thanks @TachunLin . Yes, this needs a rebase.

**Comment by torchiaf** [comment]
I just rebased the PR with upstream main branch. The `Cannot read properties of undefined (reading split)` error should be fixed now.

**Comment by TachunLin** [comment]
Thanks for the rebase with main. 

* Tested again on the staging Jenkins with a `Full` cypress test. 
  http://172.19.98.192:8080/job/harvester-cypress-test/10/console

* The `Cannot read properties of undefined (reading split)` have been diminished 

* Check the test report, we can find the "`Stop VM Negative`" and "`VM scheduling on Sepecific node`" both `failed`. 
  image
  
* It looks like the test script still trying to locate the `harvester-node-0` node in these scenarios
  image

* Check the online Harvester environment, the node name have been changed to `test-custom-name` which cause the failure
image

In order to test this pull request fix, I added the following content to the [harvester-baremetal-ansible](https://github.com/harvester/harvester-baremetal-ansible) repo.


* [ansible/roles/cypress_tests/templates/cypress.config.ts.j2](https://github.com/TachunLin/harvester-baremetal-ansible/blob/3356b4a73a92604d90e8f52080ab7595e5a8cda0/ansible/roles/cypress_tests/templates/cypress.config.ts.j2#L20-L29)


  ```
      host: {
        name: '',
        customName: '',
        disks: [
          {
            name: '',
            devPath: '',
          },
        ],
        witnessNode: false
  ```

* [ansible/roles/cypress_tests/templates/cypress.env.json.j2](https://github.com/TachunLin/harvester-baremetal-ansible/blob/3356b4a73a92604d90e8f52080ab7595e5a8cda0/ansible/roles/cypress_tests/templates/cypress.env.json.j2#L12-L26)

  ```
    "host": [
    {% for node_index in range(3) %}
      {
        "name": "harvester-node-{{ node_index }}",
        "customName": "",
        "disks": [
          {
            "name": "/dev/vda",
            "devPath": "/dev/vda"
          }
        ],
        "witnessNode": false
      }{% if not loop.last %},{% endif %}
    {% endfor %}
    ],
  ```
 
 @torchiaf , 
 Could you help to check whether my changes to the harvester-baremetal-anisble Jenkins config can meet the requirement and correct. Or if I miss out anything. Thank you.

**Comment by torchiaf** [comment]
In this scenario, the result of ansible/roles/cypress_tests/templates/cypress.env.json.j2 template, should be:

```
  "host": [
    {
      "name": "harvester-node-0",
      "customName": "test-custom-name",
      "disks": [
        {
          "name": "/dev/vda",
          "devPath": "/dev/vda"
        }
      ],
      "witnessNode": false
    },
    {
      "name": "harvester-node-1",
      "customName": "",
      "disks": [
        {
          "name": "/dev/vda",
          "devPath": "/dev/vda"
        }
      ],
      "witnessNode": false
    },
    {
      "name": "harvester-node-2",
      "customName": "",
      "disks": [
        {
          "name": "/dev/vda",
          "devPath": "/dev/vda"
        }
      ],
      "witnessNode": false
    }
  ],
```

I would change the template in something like, or by import a variable from where we define the host custom name and use it here:

```
  "host": [
  {% for node_index in range(3) %}
    {
      "name": "harvester-node-{{ node_index }}",
      "customName": "{% if node_index == 1 %}test-custom-name{% else %}{% endif %}",
      "disks": [
        {
          "name": "/dev/vda",
          "devPath": "/dev/vda"
        }
      ],
      "witnessNode": false
    }{% if not loop.last %},{% endif %}
  {% endfor %}
  ],
  ```

**Comment by TachunLin** [comment]
Thanks @torchiaf for your suggestion, I updated the harvester-baremetal-ansible commit pull request.
https://github.com/harvester/harvester-baremetal-ansible/pull/58

And trigger a new `full` cypress test on the staging Jenkins
http://172.19.98.192:8080/job/harvester-cypress-test/11/console

On the cypress test report: 
https://minio.provo.rancherlabs.com:31524/cypress-test-report/cypress/results/20240618-100419-6226eb6/index.html

We can find the `Stop VM Negative` and `VM scheduling on Specific node` cases are both `SUCCESS` 
image

**Review by TachunLin — APPROVED** [review]
LGTM. 
Tested can PASS the `Stop VM Negative` and `VM scheduling on Specific node` on the stating Jenkins full cypress test. 
https://minio.provo.rancherlabs.com:31524/cypress-test-report/cypress/results/20240618-100419-6226eb6/index.html

Thanks for fixing the issues.

**Pull request merged**